collect_app/src/androidTest/java/org/odk/collect/android/FormNavigationTestCase.java

```diff
@@ -0,0 +1,139 @@
+                {"fieldListInFieldList.xml", new String[] {"-1, ", "0, ", "-1, ", "0, "}},
+                {"regularGroupWithFieldListGroupInside.xml", new String[] {"-1, ", "0, 0, ", "-1, ", "0, 0, "}},
+                {"twoNestedRegularGroups.xml", new String[] {"-1, ", "0, 0, 0, ", "0, 0, 1, ", "0, 0, 2, ", "-1, ", "0, 0, 2, "}},
+                {"regularGroupWithQuestionAndRegularGroupInside.xml", new String[] {"-1, ", "0, 0, ", "0, 1, 0, ", "0, 1, 1, ", "-1, ", "0, 1, 1, "}},
+                {"regularGroupWithQuestionsAndRegularGroupInside.xml", new String[] {"-1, ", "0, 0, ", "0, 1, 0, ", "0, 2, ", "-1, ", "0, 2, "}},
+                {"fieldListWithQuestionAndRegularGroupInside.xml", new String[] {"-1, ", "0, ", "-1, ", "0, "}},
+                {"fieldListWithQuestionsAndRegularGroupsInside.xml", new String[] {"-1, ", "0, ", "-1, ", "0, "}},
+                {"threeNestedFieldListGroups.xml", new String[] {"-1, ", "0, ", "-1, ", "0, "}}
+        });
+    }
+
+    private String formName;
+    private String[] expectedIndices;
+
+    public FormNavigationTestCase(String formName, String[] expectedIndices) {
+        this.formName = formName;
+        this.expectedIndices = expectedIndices;
+    }
+
+    @Test
+    public void formNavigationTestCase() {
+        testIndices(formName, expectedIndices);
+    }
+
+    private void testIndices(String formName, String[] expectedIndices) {
+        try {
+            prepareFile(formName);
+        } catch (IOException e) {
+            Timber.i(e);
+        }
+        FormLoaderTask formLoaderTask = new FormLoaderTask(formPath(formName), null, null);
+        formLoaderTask.setFormLoaderListener(new FormLoaderListener() {
+            @Override
+            public void loadingComplete(FormLoaderTask task) {
+                try {
+                    for (int i = 0; i < expectedIndices.length - 1; i++) {
+                        FormController formController = task.getFormController();
+                        assertEquals(expectedIndices[i], formController.getFormIndex().toString());
+                        if (i < expectedIndices.length - 2) {
+                            formController.stepToNextScreenEvent();
+                        } else {
+                            formController.stepToPreviousScreenEvent();
+                        }
+                        assertEquals(expectedIndices[i + 1], formController.getFormIndex().toString());
+                    }
+                } catch (Exception e) {
+                    Timber.i(e);
+                }
+            }
+
+            @Override
+            public void loadingError(String errorMsg) {
+            }
+
+            @Override
+            public void onProgressStep(String stepMessage) {
+
+            }
+        });
+        formLoaderTask.execute(formPath(formName));
+    }
+
+    private void prepareFile(String formName) throws IOException {
```

Comment:
Pardon my Android ignorance, but why must we copy these files? Perhaps the method name, or a comment, could say why.

Reply:
It's because FormLoaderTask loads forms from SD card so we need to put all forms there. I added a comment.